Kernel/Time/TimeManagement.cpp

```diff
@@ -0,0 +1,249 @@
+    Vector<size_t> non_periodic_timers_indexes;
+    non_periodic_timers_indexes.ensure_capacity(m_hardware_timers.size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        if (!m_hardware_timers[index].is_null())
+            if (!m_hardware_timers[index]->is_periodic_capable())
+                non_periodic_timers_indexes.append(index);
+    }
+    return non_periodic_timers_indexes;
+}
+
+bool TimeManagement::is_hpet_periodic_mode_allowed()
+{
+    if (!KParams::the().has("hpet")) {
+        return true;
+    }
+    auto hpet_mode = KParams::the().get("hpet");
+    if (hpet_mode == "periodic")
+        return true;
+    if (hpet_mode == "nonperiodic")
+        return false;
+    ASSERT_NOT_REACHED();
+}
+
+bool TimeManagement::probe_and_set_non_legacy_hardware_timers()
+{
+    if (!ACPI::Parser::the().is_operable())
+        return false;
+    if (!HPET::test_and_initialize())
+        return false;
+    if (!HPET::the().comparators().size()) {
+        dbg() << "HPET initialization aborted.";
+        return false;
+    }
+    dbg() << "HPET: Setting appropriate functions to timers.";
+
+    m_hardware_timers.resize(HPET::the().comparators().size());
+    for (size_t index = 0; index < m_hardware_timers.size(); index++) {
+        m_hardware_timers[index] = HPET::the().comparators()[index];
+#ifdef TIME_DEBUG
+        dbg() << m_hardware_timers[index].ptr() << " <- " << HPET::the().comparators()[index].ptr();
+#endif
+    }
+
+    auto periodic_timer_indexes = scan_and_set_periodic_timers();
+    auto non_periodic_timer_indexes = scan_for_non_periodic_timers();
+
+    if (is_hpet_periodic_mode_allowed())
+        ASSERT(!periodic_timer_indexes.is_empty());
+
+    if (!non_periodic_timer_indexes.is_empty())
+        m_system_timer = m_hardware_timers[non_periodic_timer_indexes[0]];
+    else
+        m_system_timer = m_hardware_timers[periodic_timer_indexes[1]];
+
+    m_time_keeper_timer = m_hardware_timers[periodic_timer_indexes[0]];
+
+    m_system_timer->change_function([&](const RegisterState& regs) { update_scheduler_ticks(regs); });
+    m_system_timer->try_to_set_frequency(OPTIMAL_TICKS_PER_SECOND_RATE);
+    m_time_keeper_timer->change_function([&](const RegisterState& regs) { update_time(regs); });
+    m_time_keeper_timer->try_to_set_frequency(OPTIMAL_TICKS_PER_SECOND_RATE);
+
+    return true;
+}
+
+bool TimeManagement::probe_and_set_legacy_hardware_timers()
+{
+    if (ACPI::Parser::the().is_operable()) {
+        if (ACPI::Parser::the().x86_specific_flags().cmos_rtc_not_present) {
+            dbg() << "ACPI: CMOS RTC Not Present";
+            return false;
+        } else {
+            dbg() << "ACPI: CMOS RTC Present";
+        }
+    }
+
+    m_hardware_timers[0] = PIT::initialize([&](const RegisterState& regs) { update_scheduler_ticks(regs); });
+    m_hardware_timers[1] = RealTimeClock::create([&](const RegisterState& regs) { update_time(regs); });
```

Comment:
This code reads as if we're still using the 1024 ticks per second RTC to update wall time, which means that 2.3% of the time, `clock_gettime()` will return invalid values. Is this accounted for somewhere else? Not sure if you've explained this somewhere..